torch/csrc/jit/codegen/cuda/scheduler/registry.cpp

```diff
@@ -507,43 +507,57 @@ void SchedulerRuntimeInfo::collectVectorizationInfo(
 size_t SchedulerRuntimeInfo::collectMaxVectorizeSize(
     const at::Tensor& tensor,
     size_t max_vector_size_in_byte) {
+  const size_t max_vector_size = max_vector_size_in_byte / tensor.itemsize();
+
+  // If all dimensions are size 1 with stride 1, do not impose any
+  // restriction on vector size.
+  const bool all_dim_size_one = std::all_of(
+                                    tensor.sizes().begin(),
+                                    tensor.sizes().end(),
+                                    [](const auto size) {
+                                      return size == 1;
+                                    }) &&
+      std::all_of(tensor.strides().begin(),
+                  tensor.strides().end(),
+                  [](const auto stride) { return stride == 1; });
+
+  if (all_dim_size_one) {
+    return max_vector_size;
+  }
+
+  // If the innermost stride is not 1, no vectorization is allowed.
+  if (tensor.stride(-1) != 1) {
+    return 1;
+  }
+
+  const auto innermost_size = tensor.size(tensor.ndimension() - 1);
+
   size_t vector_size = 1;
-  size_t next_vector_size = 2;
-  bool next_size_compatible = true;
-
-  while (next_size_compatible &&
-         next_vector_size * tensor.itemsize() <= max_vector_size_in_byte) {
-    // If inner most dimension size is not divisible by new word size
-    //  then we cannot vectorize with this width. But we do not
-    //  care if all dimensions of this tensor is 1, i.e.
-    //  input is actually a un-squeezed 0-dim tensor.
-    for (size_t i = tensor.ndimension(); i > 0; i--) {
-      if (tensor.size(i - 1) != 1) {
-        if (tensor.size(tensor.ndimension() - 1) % next_vector_size != 0 ||
-            tensor.stride(tensor.ndimension() - 1) != 1) {
-          next_size_compatible = false;
-        }
-        break;
-      }
-    }
+  size_t next_vector_size = vector_size * 2;
 
-    if (!next_size_compatible) {
+  // Try until vector size exceeds the max allowed size
+  while (next_vector_size <= max_vector_size) {
+    if (innermost_size % next_vector_size != 0) {
       break;
     }
 
     // If any stride is not divisible by the next word size,
-    //  we cannot vectorize with this width.
-    for (auto stride : tensor.strides()) {
-      if (stride != 1 && stride % next_vector_size != 0) {
-        next_size_compatible = false;
+    //  we cannot vectorize with this width. Note that the innermost
+    //  stride is already validated to be 1.
+    bool stride_validated = true;
+    for (const auto i : c10::irange(tensor.ndimension() - 1)) {
+      if (tensor.strides().at(i) % next_vector_size != 0) {
+        stride_validated = false;
         break;
       }
     }
 
-    if (next_size_compatible) {
-      vector_size = next_vector_size;
-      next_vector_size *= 2;
+    if (!stride_validated) {
+      break;
     }
+
+    vector_size = next_vector_size;
+    next_vector_size *= 2;
```

Comment:
Just refactored.

Reply:
I think we could more aggressively ignore broadcast strides, but logic looks as good as it was. Maybe would be worth adding a `TODO:` mentioning we could improve it further with broadcasting

Reply:
Done